### PR TITLE
implement Any, All and Contains

### PR DESCRIPTION
**Pull request by rogpeppe**
As suggested in issue #41.

### CODE REVIEW

File: iter.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package quicktest
```
Comment (frankban) [inline iter.go]: License header.

File: mapiter_go1.11.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// +build !go1.12
```
Comment (frankban) [inline mapiter_go1.11.go]: License header (please also in race_test.go, we forgot it there).

File: mapiter.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"fmt"
+	"reflect"
+)
+
+func newMapIter(v reflect.Value) containerIter {
+	return mapIter{v.MapRange()}
+}
+
+type mapIter struct {
```
Comment (frankban) [inline mapiter.go]: mapItem implements containerItem by...

File: checker.go [hunk 4]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+func (c *anyChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
+	iter, err := newIter(got)
+	if err != nil {
+		return BadCheckf("%v", err)
+	}
+	for iter.next() {
+		err := c.elemChecker.Check(
+			iter.value().Interface(),
+			args,
+			func(key string, value interface{}) {},
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: I thought about this quite a bit. If an Any check fails, it means that all the elements failed the check. Would you print the notes from all of them, or... what?

[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: I thought about this quite a bit. If an Any check fails, it means that all the elements failed the check. Would you print the notes from all of them, or... what?

File: checker_test.go [hunk 5]
```diff
@@ -1888,6 +1888,246 @@ got args:
...
+want:
+  "a"
+`,
+}, {
+	about:   "Contains with map",
+	checker: qt.Contains,
+	// Note: we can't use more than one element here because
+	// pretty.Print output is non-deterministic.
+	// https://github.com/kr/pretty/issues/47
+	got:  map[string]string{"a": "d"},
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker_test.go]: I think this API works well as a generalisation across maps, slices and arrays. Perhaps we might add AllKeys and AnyKey later? (they'd only work on maps)

File: checker.go [hunk 6]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+// or string contains a value. It's the same as using
+// Any(Equals), except that it has a special case
+// for strings - if the first argument is a string,
+// the second argument must also be a string
+// and strings.Contains will be used.
+//
+// For example:
+//
+//	c.Assert("hello world", qt.Contains, "world")
+//	c.Assert([]int{3,5,7,99}, qt.Any, qt.Equals, 7)
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: done

File: checker.go [hunk 7]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+var Contains Checker = &containsChecker{
+	argNames: []string{"got", "want"},
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: Me neither. I don't like "submatch" and I can't think of anything better for now, so I think I'll leave it unless you have a better suggestion.

File: checker.go [hunk 8]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+}
+
+type containsChecker struct {
+	argNames
+}
+
+func (c *containsChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: done

File: checker.go [hunk 9]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+}
+
+// Any returns a Checker that uses the given checker to check elements
+// of a map, slice or array. It succeeds if any element passes the
+// check.
+//
+// For example:
+//
+//	c.Assert([]int{3,5,7,99}, qt.Any(qt.Equals), 7)
+//	c.Assert([][]string{{"a", "b"}, {"c", "d"}, qt.Any(qt.DeepEquals), []string{"c", "d"})
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: good catch!

File: checker.go [hunk 10]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+		elemChecker: c,
+	}
+}
+
+type anyChecker struct {
+	argNames
+	elemChecker Checker
+}
+
+func (c *anyChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: done

File: checker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: If there are zero or one elements, I guess we could do better, but... I think let's leave it for now and see how it feels in practice.

File: checker.go [hunk 11]
```diff
@@ -377,6 +378,144 @@ func (c *notChecker) Check(got interface{}, args []interface{}, note func(key st
...
+		elemChecker: c,
+	}
+}
+
+type allChecker struct {
+	argNames
+	elemChecker Checker
+}
+
+func (c *allChecker) Check(got interface{}, args []interface{}, notef func(key string, value interface{})) error {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: done

File: checker_test.go [hunk 12]
```diff
@@ -1888,6 +1888,246 @@ got args:
...
+	got:     []string{"hello"},
+	args:    []interface{}{5},
+	expectedCheckFailure: `
+error:
+  bad check: at index 0: bad check: regexp is not a string
+`,
+	expectedNegateFailure: `
+error:
+  bad check: at index 0: bad check: regexp is not a string
+`,
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker_test.go]: done

File: iter.go (hunk 1 above)
Comment (rogpeppe) [reply iter.go]: done

File: iter.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"fmt"
+	"reflect"
+)
+
+type containerIter interface {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply iter.go]: done

File: iter.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+		return &sliceIter{
+			index: -1,
+			v:     v,
+		}, nil
+	default:
+		return nil, fmt.Errorf("map, slice or array required")
+	}
+}
+
+type sliceIter struct {
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply iter.go]: done

File: mapiter.go [hunk 15]
```diff
@@ -0,0 +1,28 @@
+// +build go1.12
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply mapiter.go]: done

File: mapiter.go (hunk 3 above)
Comment (rogpeppe) [reply mapiter.go]: done

File: mapiter_go1.11.go (hunk 2 above)
Comment (rogpeppe) [reply mapiter_go1.11.go]: done

File: checker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (frankban) [reply checker.go]: I agree, but please let's add a comment about these concerns.

[the remark this answers is not in the document]
Comment (rogpeppe) [reply checker.go]: done

**Pull request merged**